nucliadb_node/nucliadb_node/indexer.py

```diff
@@ -0,0 +1,341 @@
+                    "Indexer task raised an error while finalizing it", exc_info=exc
+                )
+        self.indexers.clear()
+
+    async def index_message_soon(self, msg: Msg):
+        work = WorkUnit.from_msg(msg)
+        await work.mpu.start()
+
+        logger.info(
+            f"Index message for shard {work.shard_id} is being enqueued",
+            extra={
+                "shard": work.shard_id,
+                "subject": msg.subject,
+                "reply": msg.reply,
+                "seqid": work.seqid,
+                "storage_key": work.index_message.storage_key,
+            },
+        )
+
+        await self.push_work_to_indexer(work)
+        CONCURRENT_INDEXERS_COUNT.set(
+            len(self.indexers), labels=dict(node=work.node_id)
+        )
+
+    async def push_work_to_indexer(self, work: WorkUnit):
+        exists = work.shard_id in self.indexers
+        if exists:
+            indexer, _ = self.indexers[work.shard_id]
+            indexer.index_soon(work)
+        else:
+            indexer = PriorityIndexer(writer=self.writer, storage=self.storage)
+            indexer.index_soon(work)
+            task = asyncio.create_task(indexer.work_until_finish())
+            task.add_done_callback(
+                partial(
+                    ConcurrentShardIndexer._indexer_done_callback, self, work.shard_id
+                )
+            )
+            self.indexers[work.shard_id] = (indexer, task)
+
+    def _indexer_done_callback(self, shard_id: str, task: asyncio.Task):
+        # autoremove indexer when finished
+        self.indexers.pop(shard_id)
+
+        # propagate errors
+        if task.exception() is not None:
+            raise task.exception()  # type: ignore
+
+
+class PriorityIndexer:
+    def __init__(self, writer: Writer, storage: Storage):
+        self.writer = writer
+        self.storage = storage
+        self.work_queue: asyncio.PriorityQueue = asyncio.PriorityQueue()
+        self.working = False
+
+    def index_soon(self, work: WorkUnit):
+        self.work_queue.put_nowait(work)
+
+    async def work_until_finish(self):
+        self.working = True
+        try:
+            while not self.work_queue.empty():
+                work = await self.work_queue.get()
+                await self._do_work(work)
+        except Exception as exc:
+            # if an exception occurred, we can't longer ensure proper ordering
+            # for this queue. We flush the queue and wait for the messages to be
+            # redelivered
+            event_id = capture_exception(exc)
+            logger.error(
+                "An error happened on an indexer, all messages for its queue will be flushed. "
+                f"Check sentry for more details. Event id: {event_id}",
+                extra={
+                    "seqid": work.seqid,
+                    "shard": work.index_message.shard,
+                    "storage_key": work.index_message.storage_key,
```

Comment:
is the kbid part of the storage key? if not, it would be nice to have the kbid in the log extra too.